asreview/webapp/authentication/models.py

```diff
@@ -0,0 +1,272 @@
+        email=None,
+        name=None,
+        affiliation=None,
+        password=None,
+        confirmed=False,
+        public=True,
+    ):
+
+        self.identifier = identifier
+        self.origin = origin
+        self.email = email
+        self.name = name
+        self.affiliation = affiliation
+        if bool(password):
+            self.hashed_password = generate_password_hash(password)
+        self.confirmed = confirmed
+        self.public = public
+
+    def update_profile(self, email, name, affiliation, password=None, public=True):
+
+        self.email = email
+        self.name = name
+        self.affiliation = affiliation
+        if self.origin == "asreview" and bool(password):
+            self.hashed_password = generate_password_hash(password)
+        self.public = public
+
+        return self
+
+    def reset_password(self, new_password):
+        if self.origin == "asreview" and bool(new_password):
+            self.hashed_password = generate_password_hash(new_password)
+        # reset token
+        self.token = None
+        self.token_created_at = None
+        return self
+
+    def set_token_data(self, secret, salt):
+        """Set token data (used in email verification after
+        init, and for forgot-password"""
+        token, token_created_at = User.generate_token_data(secret, salt, self.email)
+        self.token = token
+        self.token_created_at = token_created_at
+        return self
+
+    def verify_password(self, password):
+        """Verify password"""
+        if bool(self.hashed_password):
+            return check_password_hash(self.hashed_password, password)
+        else:
+            return False
+
+    def get_name(self):
+        """Get name-ish thing from user account"""
+        name = self.name or self.email
+        return name
+
+    def summarize(self):
+        """Summarize user account in frontend data packet"""
+        return {"id": self.id, "name": self.get_name(), "email": self.email}
+
+    def confirm_user(self):
+        """This function confirms a user by setting the confirmed
+        field to True and removes the token data"""
+        self.confirmed = True
+        self.token = None
+        self.token_created_at = None
+        return self
+
+    def token_valid(self, provided_token, max_hours=24):
+        """Checks whether provided token is correct and still valid"""
+        # there must be a token and a timestamp
+        if bool(self.token) and bool(self.token_created_at):
+            # what is now
+            now = dt.datetime.utcnow()
+            # get time-difference in hours
+            diff = now - self.token_created_at
+            # give me hours and remaining seconds
+            [hours, r_secs] = divmod(diff.total_seconds(), 3600)
+            # return if token is correct and we are still before deadline
+            return self.token == provided_token and (
+                hours <= max_hours or (hours == max_hours and r_secs < 60)
```

Comment:
So the token is valid for max_hours + 60 seconds if I understand it correctly? Any reason why it's like this? It seems a bit complicated.